solr/core/src/java/org/apache/solr/search/grouping/distributed/responseprocessor/SkipSecondStepSearchGroupShardResponseProcessor.java

```diff
@@ -0,0 +1,118 @@
+import org.apache.lucene.search.grouping.SearchGroup;
+import org.apache.lucene.search.grouping.TopGroups;
+import org.apache.lucene.util.BytesRef;
+import org.apache.solr.handler.component.ResponseBuilder;
+import org.apache.solr.handler.component.ShardDoc;
+import org.apache.solr.handler.component.ShardResponse;
+import org.apache.solr.search.SolrIndexSearcher;
+import org.apache.solr.search.grouping.GroupingSpecification;
+import org.apache.solr.search.grouping.distributed.shardresultserializer.SearchGroupsResultTransformer;
+
+public class SkipSecondStepSearchGroupShardResponseProcessor extends SearchGroupShardResponseProcessor {
+
+  @Override
+  protected SearchGroupsResultTransformer newSearchGroupsResultTransformer(SolrIndexSearcher solrIndexSearcher) {
+    return new SearchGroupsResultTransformer.SkipSecondStepSearchResultResultTransformer(solrIndexSearcher);
+  }
+
+  @Override
+  protected SearchGroupsContainer newSearchGroupsContainer(ResponseBuilder rb) {
+    return new SkipSecondStepSearchGroupsContainer(rb.getGroupingSpec().getFields());
+  }
+
+  protected static class SkipSecondStepSearchGroupsContainer extends SearchGroupsContainer {
+
+    private final Map<Object, String> docIdToShard = new HashMap<>();
+
+    public SkipSecondStepSearchGroupsContainer(String[] fields) {
+      super(fields);
+    }
+
+    @Override
+    public void addSearchGroups(ShardResponse srsp, String field, Collection<SearchGroup<BytesRef>> searchGroups) {
+      super.addSearchGroups(srsp, field, searchGroups);
+      for (SearchGroup<BytesRef> searchGroup : searchGroups) {
+        assert(srsp.getShard() != null);
+        docIdToShard.put(searchGroup.topDocSolrId, srsp.getShard());
+      }
+    }
+
+    @Override
+    public void addMergedSearchGroups(ResponseBuilder rb, String groupField, Collection<SearchGroup<BytesRef>> mergedTopGroups ) {
+      // TODO: add comment or javadoc re: why this method is overridden as a no-op
+    }
+
+    @Override
+    public void addSearchGroupToShards(ResponseBuilder rb, String groupField, Collection<SearchGroup<BytesRef>> mergedTopGroups) {
+      super.addSearchGroupToShards(rb, groupField, mergedTopGroups);
+
+      final GroupingSpecification groupingSpecification = rb.getGroupingSpec();
+      final Sort groupSort = groupingSpecification.getGroupSort();
+      final String[] fields = groupingSpecification.getFields();
+
+      GroupDocs<BytesRef>[] groups = new GroupDocs[mergedTopGroups.size()];
+      Map<Object, ShardDoc> resultsId = new HashMap<>(mergedTopGroups.size());
+
+      // This is the max score found in any document on any group
+      float maxScore = 0;
+      int index = 0;
+
+      for (SearchGroup<BytesRef> group : mergedTopGroups) {
+        maxScore = Math.max(maxScore, group.topDocScore);
+        final String shard = docIdToShard.get(group.topDocSolrId);
+        assert(shard != null);
+        ShardDoc sdoc = new ShardDoc(group.topDocScore,
+            fields,
```

Comment:
The passing of `fields` to the `ShardDoc` here seems incorrect to me:
* `ShardDoc` is a `FieldDoc` and per https://github.com/apache/lucene-solr/blob/releases/lucene-solr/8.0.0/lucene/core/src/java/org/apache/lucene/search/FieldDoc.java#L43-L51 it expects "... values which are used to sort ..." to be stored in that field (e.g. 2019).
* Instead here `GroupingSpecification.getFields()` is passed (e.g. `publicationYear`) i.e. a field name rather than a field value.

On the `group.skip.second.step=true` code path that particular argument is not actually used in practice, hence this name-vs-value difference can easily go unnoticed. Something like commit https://github.com/cpoerschke/lucene-solr/commit/2b3069b52ac514478e4f6363d18b485a93eae566 would be a way to use the other `ShardDoc` constructor.

Reply:
Thanks, added in 72f559a0